llvm/include/llvm/CodeGen/TargetLowering.h

```diff
@@ -3334,6 +3334,10 @@ class TargetLoweringBase {
     return false;
   }
 
+  /// Try to convert an extract element of a vector setcc operation into an
+  /// extract element followed by a scalar operation.
+  virtual bool shouldScalarizeSetCC(SDValue VecOp) const { return false; }
```

Comment:
I don't see why this needs a special case hook. We already do this for extractelement other_op for the hasOneUse case

Reply:
I was simply following the existing pattern for shouldScalarizeBinOp as it looks like not all targets prefer to do this type of optimisation. I did try enabling this for all targets and it seems to give regressions on X86. Or are you referring to something else?

Reply:
I could just reuse the existing `shouldScalarizeBinOp` hook, but the setcc op feels like a special case because the result type is often different to the input (e.g. v4i1 output, v4i32 input, etc.) and may set condition registers.

Reply:
Hmm, perhaps I didn't understand your comment. Are you implying that I can simply reuse `shouldScalarizeBinOp` and check the opcode of `VecOp` instead of adding an extra hook just for setcc? If so, then I'm happy to do that - it just requires modifying all existing `shouldScalarizeBinOp` hooks to bail out for the setcc case.

Reply:
I haven't looked at the context that closely, but we need way fewer points of configuration. The ones we do have should be much more general. This stuff is not discoverable by backend maintainers, and is high effort to find and implement. But consider this a hard -1 on adding a new super specific thing. If a target hook as no arguments, it's a major code smell.

But yes, shouldScalarizeBinOp would be preferable. The changing result type probably doesn't matter in practice, it's approximately fixed per target

Reply:
Done